cmd/snap/cmd_download.go

```diff
@@ -97,20 +130,187 @@ func fetchSnapAssertions(tsto *image.ToolingStore, snapPath string, snapInfo *sn
 	return assertPath, err
 }
 
-func printInstallHint(assertPath, snapPath string) {
-	// simplify paths
-	wd, _ := os.Getwd()
-	if p, err := filepath.Rel(wd, assertPath); err == nil {
-		assertPath = p
+//
+type downloadStore interface {
+	DownloadSnap(name string, opts image.DownloadOptions) (targetFn string, info *snap.Info, err error)
+	AssertionFetcher(db *asserts.Database, save func(asserts.Assertion) error) asserts.Fetcher
+}
+
+var newDownloadStore = func() (downloadStore, error) {
+	return image.NewToolingStore()
+}
+
+func (x *cmdDownload) downloadDirect(snapName string, revision snap.Revision) error {
+	tsto, err := newDownloadStore()
+	if err != nil {
+		return err
 	}
-	if p, err := filepath.Rel(wd, snapPath); err == nil {
-		snapPath = p
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching snap %q\n"), snapName)
+	dlOpts := image.DownloadOptions{
+		TargetDir: x.TargetDir,
+		Basename:  x.Basename,
+		Channel:   x.Channel,
+		CohortKey: x.CohortKey,
+		Revision:  revision,
+		// if something goes wrong, don't force it to start over again
+		LeavePartialOnError: true,
 	}
-	// add a hint what to do with the downloaded snap (LP:1676707)
-	fmt.Fprintf(Stdout, i18n.G(`Install the snap with:
-   snap ack %s
-   snap install %s
-`), assertPath, snapPath)
+	snapPath, snapInfo, err := tsto.DownloadSnap(snapName, dlOpts)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(Stdout, i18n.G("Fetching assertions for %q\n"), snapName)
+	assertPath, err := fetchSnapAssertionsDirect(tsto, snapPath, snapInfo)
+	if err != nil {
+		return err
+	}
+	printInstallHint(assertPath, snapPath)
+	return nil
+}
+
+func downloadSnapFromStreamWithProgress(downloadPath string, stream io.ReadCloser, dlInfo *client.DownloadInfo) error {
+	// TODO: support resume of exiting files and not download if the
+	//       file is already there with the right hash
+	f, err := os.OpenFile(downloadPath, os.O_RDWR|os.O_CREATE, 0600)
```

Comment:
As long as the TODO is present don't we need to use `os.O_TRUNC` for correct behaviour in case of unexpected resume-like situation?

Reply:
probably, yes. If the end result is wrong the checksumming will catch it, so not a huge issue methinks.